### ISSUE TITLE
Fatal error: Uncaught TypeError: Argument 1 passed to unserialize() must be an instance of string, null given

### ISSUE THREAD

**Issue by donutloop** [facebook/hhvm#7178]
HHVM Version: 3.14.1 

Hello,
i loaded the binary (Latest Snapshot and 1.1.2) with wget. Then i tried to install with "hhvm --php composer.phar install" the Dependencies. Now i get these error:

#Fatal error: Uncaught TypeError: Argument 1 passed to unserialize() must be an instance of string, null given in /var/www/app/composer.phar:23
Stack trace:
#0 (): unserialize()
#1 (): __SystemLib\PharArchiveHandler->parsePhar()
#2 (): __SystemLib\PharArchiveHandler->__construct()
#3 (): Phar->__construct()
#4 /var/www/blog-api/composer.phar(23): Phar::mapPhar()
#5 {main}

Composer 
 #7171 

Best regards,
Marcel

**Comment by Orvid** [comment]
@nazar-pc Could this be related in-part to the small refactor that was done while adding support for compressed Phars?

**Comment by nazar-pc** [comment]
Theoretically it can be the case, because significant amount of code was refactored, but I'm not sure how exactly.

@donutloop, I've tried latest nightly build on Ubuntu 16.10 and I do not see this error on my machine.
Can you provide your `composer.json` and distro/version so that it is possible to reproduce exactly the same test case?

**Comment by Orvid** [comment]
We have a nightly on 16.10? O_o....

**Comment by donutloop** [comment]
PHP version (7.0.99-hhvm)
Debian version: jessie

**Comment by donutloop** [comment]
composer.json

`{
  "name": "",
  "license": "MIT",
  "type": "project",
  "description": "",
  "autoload": {
    "psr-4": { "": "src/" },
    "classmap": [ "app/AppKernel.php", "app/AppCache.php" ]
  },
  "autoload-dev": {
    "psr-4": { "Tests\\": "tests/" }
  },
  "require": {
    "php": ">=7.0",
    "symfony/symfony": "3.1.*",
    "doctrine/orm": "^2.5",
    "doctrine/doctrine-bundle": "^1.6",
    "doctrine/doctrine-cache-bundle": "^1.2",
    "symfony/swiftmailer-bundle": "^2.3",
    "symfony/monolog-bundle": "^2.8",
    "sensio/distribution-bundle": "^5.0",
    "sensio/framework-extra-bundle": "^3.0.2",
    "nelmio/api-doc-bundle": "^2.12",
    "friendsofsymfony/user-bundle": "~2.0@dev",
    "friendsofsymfony/oauth-server-bundle": "^1.5",
    "jms/serializer-bundle": "^1.1",
    "friendsofsymfony/rest-bundle": "^1.7",
    "willdurand/hateoas-bundle": "^1.1"
  },
  "require-dev": {
    "sensio/generator-bundle": "^3.0",
    "symfony/phpunit-bridge": "^3.0",
    "doctrine/doctrine-fixtures-bundle": "^2.3",
    "liip/functional-test-bundle": "^1.6"
  },
  "scripts": {
    "post-install-cmd": [
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
    ],
    "post-update-cmd": [
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
      "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
    ]
  },
  "config": {
    "platform": {
      "php": "7.0"
    }
  },
  "extra": {
    "symfony-app-dir": "app",
    "symfony-bin-dir": "bin",
    "symfony-var-dir": "var",
    "symfony-web-dir": "web",
    "symfony-tests-dir": "tests",
    "symfony-assets-install": "relative",
    "branch-alias": {
      "dev-master": "3.0-dev"
    }
  }
}
`

**Comment by nazar-pc** [comment]
Nope, nightly from Xenial, but I'm running it on Yakkety already (hey, I've used even Wily builds on Yakkety:))

**Comment by nazar-pc** [comment]
OK, so Xenial build of HHVM, Composer 1.1.2, I'm getting the same result as with PHP 7.0.7, namely:
```
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing ...

doctrine/data-fixtures suggests installing doctrine/mongodb-odm (For loading MongoDB ODM fixtures)
doctrine/data-fixtures suggests installing doctrine/phpcr-odm (For loading PHPCR ODM fixtures)
paragonie/random_compat suggests installing ext-libsodium (Provides a modern crypto API that can be used to generate random bytes.)
symfony/polyfill-intl-icu suggests installing ext-intl (For best performance)
doctrine/doctrine-cache-bundle suggests installing symfony/security-acl (For using this bundle to cache ACLs)
friendsofsymfony/user-bundle suggests installing willdurand/propel-typehintable-behavior (Needed when using the propel implementation)
monolog/monolog suggests installing graylog2/gelf-php (Allow sending log messages to a GrayLog2 server)
monolog/monolog suggests installing raven/raven (Allow sending log messages to a Sentry server)
monolog/monolog suggests installing doctrine/couchdb (Allow sending log messages to a CouchDB server)
monolog/monolog suggests installing ruflin/elastica (Allow sending log messages to an Elastic Search server)
monolog/monolog suggests installing php-amqplib/php-amqplib (Allow sending log messages to an AMQP server using php-amqplib)
monolog/monolog suggests installing ext-amqp (Allow sending log messages to an AMQP server (1.0+ required))
monolog/monolog suggests installing ext-mongo (Allow sending log messages to a MongoDB server)
monolog/monolog suggests installing mongodb/mongodb (Allow sending log messages to a MongoDB server via PHP Driver)
monolog/monolog suggests installing aws/aws-sdk-php (Allow sending log messages to AWS services like DynamoDB)
monolog/monolog suggests installing rollbar/rollbar (Allow sending log messages to Rollbar)
monolog/monolog suggests installing php-console/php-console (Allow sending log messages to Google Chrome)
sensio/framework-extra-bundle suggests installing symfony/psr-http-message-bridge (To use the PSR-7 converters)
nelmio/api-doc-bundle suggests installing dunglas/api-bundle (For making use of resources definitions of DunglasApiBundle.)
friendsofsymfony/oauth-server-bundle suggests installing doctrine/mongodb-odm-bundle (*)
friendsofsymfony/oauth-server-bundle suggests installing propel/propel-bundle (If you want to use Propel with Symfony2, then you will have to install the PropelBundle)
friendsofsymfony/oauth-server-bundle suggests installing willdurand/propel-typehintable-behavior (The Typehintable behavior is useful to add type hints on generated methods, to be compliant with interfaces)
jms/serializer-bundle suggests installing jms/di-extra-bundle (Required to get lazy loading (de)serialization visitors, ~1.3)
liip/functional-test-bundle suggests installing nelmio/alice (Required when using loadFixtureFiles functionality)
liip/functional-test-bundle suggests installing hautelook/alice-bundle (Required when using loadFixtureFiles functionality with custom providers)
liip/functional-test-bundle suggests installing brianium/paratest (Required when using paratest to parallelize tests)
Writing lock file
Generating autoload files

                                                                                                         
  [RuntimeException]                                                                                     
  Could not scan for classes inside "app/AppKernel.php" which does not appear to be a file nor a folder  
                                                                                                         

install [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--ignore-platform-reqs] [--] [<packages>]...
```

**Comment by donutloop** [comment]
i use a docker container. Is that maybe the issue?

**Comment by nazar-pc** [comment]
Is it public, if so - can you give a link to it?

**Comment by donutloop** [comment]
isn't public. how shall i  transferring the files?

i use the option:  hhvm.php7.all = 1

**Comment by donutloop** [comment]
My docker configuration work fine with php5

**Comment by nazar-pc** [comment]
I can reproduce an issue with `hhvm.php7.all = 1` on current system

**Comment by donutloop** [comment]
ok cool :+1:

**Comment by nazar-pc** [comment]
I've extracted `composer.phar` of version 1.1.2 with PHP:
```php
<?php
$p = new Phar(__DIR__.'/composer.phar');
$p->extractTo(__DIR__.'/composer');
```

And tried to execute it from directory in order to avoid using Phar. Without `hhvm.php7.all = 1` as well as under PHP everything is fine, but with `hhvm.php7.all = 1`:
```
> hhvm --php composer/bin/composer -vvv install
Reading ./composer.json
Loading config file /home/nazar-pc/.composer/config.json
Loading config file /home/nazar-pc/.composer/auth.json
Loading config file ./composer.json

                                                        
  [ErrorException]                                      
  rtrim() expects parameter 1 to be string, null given  
                                                        

Exception trace:
 () at /*/composer/src/Composer/Config.php:-1
 Composer\Util\ErrorHandler::handle() at n/a:n/a
 rtrim() at /*/composer/src/Composer/Config.php:-1
 Composer\Config->get() at /*/composer/src/Composer/Factory.php:-1
 Composer\Factory::createRemoteFilesystem() at /*/composer/src/Composer/Factory.php:-1
 Composer\Factory->createComposer() at /*/composer/src/Composer/Factory.php:-1
 Composer\Factory::create() at /*/composer/src/Composer/Console/Application.php:-1
 Composer\Console\Application->getComposer() at /*/composer/src/Composer/Console/Application.php:-1
 Composer\Console\Application->getPluginCommands() at /*/composer/src/Composer/Console/Application.php:-1
 Composer\Console\Application->doRun() at /*/composer/vendor/symfony/console/Application.php:-1
 Symfony\Component\Console\Application->run() at /*/composer/src/Composer/Console/Application.php:-1
 Composer\Console\Application->run() at /*/composer/bin/composer:-1
```

So this is not Phar, or at least not only Phar. Which is even worse - line number is always `-1`.

**Comment by donutloop** [comment]
How can i use php 7 (Without these option)? That i could use Composer on HHVM + PHP7 . Do you have some links ?

**Comment by nazar-pc** [comment]
You can remove that option from config and simply use Composer, it should work out of the box.
`hhvm.php7.all` effect is described in [documentation](https://docs.hhvm.com/hhvm/configuration/INI-settings#php-7-settings), I believe it shouldn't have any specific incompatibility with Composer regardless of this option being present or not, but apparently there is some not very obvious difference, so just don't set that option for now, you're not likely to hit those edge cases with carefully written PHP5 code anyway.

**Comment by apeabody** [comment]
OK, I was able to get the composer.json provided by @donutloop to run under a **fresh** Debian Jessie box (small plug for a blog article I wrote recently on using Vagrant with VirtualBox for this exact purpose: http://blog.andrewpeabody.com/2016/06/developing-locally-in-linux-boxes-using.html ).

- hhvm-nightly amd64 2016.06.29~xenial
```
sudo apt-key adv --recv-keys --keyserver hkp://keyserver.ubuntu.com:80 0x5a16e7281be7a449
echo deb http://dl.hhvm.com/debian jessie main | sudo tee /etc/apt/sources.list.d/hhvm.list
sudo apt-get update
sudo apt-get install hhvm-nightly```

- Composer 1.1.**3** (NOTE: brand new version from ~4 days ago on the 26th)
```
php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
php -r "if (hash_file('SHA384', 'composer-setup.php') === 'e115a8dc7871f15d853148a7fbac7da27d6c0030b848d9b3dc09e2a0388afed865e6a3d6b3c0fad45c48e2b5fc1196ae') { echo 'Installer verified'; } else { echo 'Installer corrupt'; unlink('composer-setup.php'); } echo PHP_EOL;"
php composer-setup.php
php -r "unlink('composer-setup.php');"
```

No special configuration other than creating the above composer.json, and then running:
`hhvm --php composer.phar install`

NOTE: All clean except for the warning about the missing proprietary/local "app/AppKernel.php" and Sensio post install/update stuff as expected.

References #7185 

Good Luck!

**Comment by aorenste** [comment]
It looks like this was fixed by #7197.  Closing.
